src/main/scala/org/broadinstitute/hail/keytable/KeyTable.scala

```diff
@@ -0,0 +1,397 @@
+package org.broadinstitute.hail.keytable
+
+import org.apache.spark.SparkContext
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import org.broadinstitute.hail.annotations._
+import org.broadinstitute.hail.check.Gen
+import org.broadinstitute.hail.expr._
+import org.broadinstitute.hail.methods.{Aggregators, Filter}
+import org.broadinstitute.hail.utils._
+import org.broadinstitute.hail.io.TextExporter
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+
+object KeyTable extends Serializable with TextExporter {
+
+  def importTextTable(sc: SparkContext, path: Array[String], keyNames: String, nPartitions: Int, config: TextTableConfiguration) = {
+    val files = sc.hadoopConfiguration.globAll(path)
+    if (files.isEmpty)
+      fatal("Arguments referred to no files")
+    
+    val keyNameArray = Parser.parseIdentifierList(keyNames)
+
+    val (struct, rdd) =
+      if (nPartitions < 1)
+        fatal("requested number of partitions in -n/--npartitions must be positive")
+      else
+        TextTableReader.read(sc)(files, config, nPartitions)
+
+    val keyNamesValid = keyNameArray.forall { k =>
+      val res = struct.selfField(k).isDefined
+      if (!res)
+        println(s"Key `$k' is not present in input table")
+      res
+    }
+    if (!keyNamesValid)
+      fatal("Invalid key names given")
+
+    KeyTable(rdd.map(_.value), struct, keyNameArray)
+  }
+
+  def annotationToSeq(a: Annotation, nFields: Int) = Option(a).map(_.asInstanceOf[Row].toSeq).getOrElse(Seq.fill[Any](nFields)(null))
+
+  def setEvalContext(ec: EvalContext, k: Annotation, v: Annotation, nKeys: Int, nValues: Int) =
+    ec.setAll(annotationToSeq(k, nKeys) ++ annotationToSeq(v, nValues): _*)
+
+  def setEvalContext(ec: EvalContext, a: Annotation, nFields: Int) =
+    ec.setAll(annotationToSeq(a, nFields): _*)
+
+  def toSingleRDD(rdd: RDD[(Annotation, Annotation)], nKeys: Int, nValues: Int): RDD[Annotation] =
+    rdd.map { case (k, v) =>
+      val x = Annotation.fromSeq(annotationToSeq(k, nKeys) ++ annotationToSeq(v, nValues))
+      x
+    }
+
+  def apply(rdd: RDD[Annotation], signature: TStruct, keyNames: Array[String]): KeyTable = {
+    val keyFields = signature.fields.filter(fd => keyNames.contains(fd.name))
+    val keyIndices = keyFields.map(_.index)
+
+    val valueFields = signature.fields.filterNot(fd => keyNames.contains(fd.name))
+    val valueIndices = valueFields.map(_.index)
+
+    assert(keyIndices.toSet.intersect(valueIndices.toSet).isEmpty)
```

Comment:
can this ever happen?  If so, should print a nice error message instead of an assertion error.  If not, continue